### ISSUE TITLE
ssl:sockname returns an error when using socket backend

### ISSUE THREAD

**Issue by chvish** [erlang/otp#5930]
**Describe the bug**
Invoking `ssl:sockname/1` on a lsocket creating using the [socket](https://www.erlang.org/doc/man/socket.html) backend returns in an error.

**To Reproduce**
```
2> {ok, S} = ssl:listen(0, [{inet_backend, socket}]), ssl:sockname(S).
** exception error: no function clause matching
                    ssl:sockname({sslsocket,nil,
                                     {{'$inet',gen_tcp_socket,
                                          {<0.111.0>,
```

**Expected behavior**
Return the expected address as per function definition

**Affected versions**
OTP 24 and OTP 25 RC

**Additional context**
There is not a lot of detail on whether the socket backend is supported for SSL that I could find, but given that SSL socket creation is supported, ssl:sockname/1 should hopefully work as well.

**Comment by IngelaAndin** [comment]
For OTP-25 it is definitely meant to work. During OTP 24 it is a work in progress, but we might include a fix in some other patch to OTP-24. I found some legacy guards that I have removed and this should solve your particular problem. We are still looking into some test issues when the new backend seems to behave differently from the old.

**Comment by chvish** [comment]
Thanks @IngelaAndin 

Is there some documentation on where the socket backend is incompatible? And would it be fair to assume that the socket backend should still be considered _experimental_?

It would also help if there is a roadmap on full compatibility for the socket backend in OTP 24.

**Comment by IngelaAndin** [comment]
Yes, it is still to be considered experimental. We have been working on that it should be working well enough on all platforms so that we can then enable it on all our tests to run with the new backend so that hopefully for OTP-26 it could be made the default and not be experimental anymore.  @bmk would you like to add anything?

**Comment by bmk** [comment]
We found two problem in gen_tcp (socket). One with the raw option and one with shutdown.
I hope these can be included in 25.0.
We do test inet_backend = socket on two *other* applications (other than kernel); megaco and snmp.
